massa-execution-worker/src/speculative_ledger.rs

```diff
@@ -175,14 +267,93 @@ impl SpeculativeLedger {
     /// Creates a new smart contract address with initial bytecode.
     ///
     /// # Arguments
+    /// * `creator_address`: address that asked for this creation. Will pay the storage costs.
     /// * `addr`: address to create
     /// * `bytecode`: bytecode to set in the new ledger entry
     pub fn create_new_sc_address(
         &mut self,
+        creator_address: Address,
         addr: Address,
         bytecode: Vec<u8>,
     ) -> Result<(), ExecutionError> {
-        // set bytecode (create if do not exist)
+        // check for address existence
+        if !self.entry_exists(&creator_address) {
+            return Err(ExecutionError::RuntimeError(format!(
+                "could not set bytecode for address {}: entry does not exist",
+                addr
+            )));
+        }
+
+        // check that we don't collide with existing address
+        if self.entry_exists(&addr) {
+            return Err(ExecutionError::RuntimeError(format!(
+                "could not set bytecode for address {}: address generated to store bytecode already exist. Try again.",
+                addr
+            )));
+        }
+
+        if bytecode.len() > self.max_bytecode_size as usize {
+            return Err(ExecutionError::RuntimeError(format!(
+                "could not set bytecode for address {}: bytecode size exceeds maximum allowed size",
+                addr
+            )));
+        }
+
+        // calculate the cost of storing the address and bytecode
+        let address_storage_cost = self
+            .storage_costs_constants
+            .ledger_cost_per_byte
+            .checked_mul_u64(
+                self.storage_costs_constants
+                    .ledger_entry_base_size
+                    .try_into()
+                    .map_err(|_| {
+                        ExecutionError::RuntimeError(
+                            "overflow in calculating cost for base entry (address + balance)"
+                                .to_string(),
+                        )
+                    })?,
+            )
+            .ok_or_else(|| {
+                ExecutionError::RuntimeError("overflow in ledger cost for balance".to_string())
+            })?
+            // Bytecode key and data
+            .checked_add(
+                self.storage_costs_constants
+                    .ledger_cost_per_byte
+                    .checked_mul_u64(ADDRESS_SIZE_BYTES.try_into().map_err(|_| {
```

Comment:
this can be removed as the address size is factored inside ledger_entry_base_size

Reply:
Bytecode has a key which address (plus an identifier that I don't count in costs)

Reply:
the size of (address + bytecode identifier) is equal to the base ledger entry size